### PR TITLE
Filtering torrents in transmission-remote

### PR DESCRIPTION
**Pull request by MatanZ**
This PR adds support for filtering torrents by various criteria to transmission-remote, per the discussion in #2948.
The possible filters are  uploading, downloading, label, torrent name, and upload ratio. But it is easy to add more.
Filters may be chained and may be negated.

The implementation is not the most efficient. In particulat it causes double calling of torrent-get RPC if this call is needed for the action that follows the filter. But I think it is does not matter for most use cases.

### CODE REVIEW

**Comment by GaryElshaw** [comment]
I just noticed the naming in 'Files changed', is it correct? transmission-remote.1

File: utils/remote.cc [hunk 1]
```diff
@@ -1974,8 +1984,128 @@ static void printGroups(tr_variant* top)
...
+                continue;
+            }
+            /*
+            if (tr_variantDictFindInt(d, TR_KEY_eta, &eta) && 
+                tr_variantDictFindInt(d, TR_KEY_leftUntilDone, &leftUntilDone) &&
+                tr_variantDictFindStrView(d, TR_KEY_name, &name) && tr_variantDictFindInt(d, TR_KEY_rateDownload, &down) &&
+                tr_variantDictFindInt(d, TR_KEY_rateUpload, &up) &&
+                tr_variantDictFindInt(d, TR_KEY_sizeWhenDone, &sizeWhenDone) &&
+                && tr_variantDictFindReal(d, TR_KEY_uploadRatio, &ratio))
+                */
```
Comment (ckerr) [inline utils/remote.cc:2026]: Is this needed?

File: utils/remote.cc [hunk 2]
```diff
@@ -1974,8 +1984,128 @@ static void printGroups(tr_variant* top)
...
+            bool include = negate;
+            std::string status = getStatusString(d);
```
Comment (ckerr) [inline utils/remote.cc:2028]:
```suggestion
            auto const status = getStatusString(d);
```

File: utils/remote.cc [hunk 3]
```diff
@@ -1974,8 +1984,128 @@ static void printGroups(tr_variant* top)
...
+                double ratio;
+                if (!tr_variantDictFindReal(d, TR_KEY_uploadRatio, &ratio))
+                {
+                    continue;
+                }
+                if (ratio >= std::stof(std::string(arg)))
+                {
+                    include = !include;
+                }
+                break;
```
Comment (ckerr) [inline utils/remote.cc:2090]: I'm surprised that the compiler lets you declare a local like this after the case statement...? I didn't think that was legal.

Anyway, the scope of `ratio` can be reduced here by using an if decl block:

```suggestion
                double ratio;
                if (double ratio = {}; !tr_variantDictFindReal(d, TR_KEY_uploadRatio, &ratio))
                {
                    continue;
                }
                else if (ratio >= std::stof(std::string(arg)))
                {
                    include = !include;
                }
                break;
```

File: utils/remote.cc [hunk 4]
```diff
@@ -1974,8 +1984,128 @@ static void printGroups(tr_variant* top)
...
+            case 'u': // Uploading (Status is Uploading, Up&Down or Seeding
+                if ((status.find("Up") != std::string::npos) || (status == "Seeding"))
+                {
+                    include = !include;
+                }
+                break;
+            case 'l': // label
+                tr_variant* l;
+                if (tr_variantDictFindList(d, TR_KEY_labels, &l))
```
Comment (ckerr) [inline utils/remote.cc:2052]: Scope of `l` can be narrowed

```suggestion
                if (tr_variant* l = nullptr; tr_variantDictFindList(d, TR_KEY_labels, &l))
```

File: utils/remote.cc [hunk 5]
```diff
@@ -1974,8 +1984,128 @@ static void printGroups(tr_variant* top)
...
+            if (tr_variantDictFindInt(d, TR_KEY_eta, &eta) && 
+                tr_variantDictFindInt(d, TR_KEY_leftUntilDone, &leftUntilDone) &&
+                tr_variantDictFindStrView(d, TR_KEY_name, &name) && tr_variantDictFindInt(d, TR_KEY_rateDownload, &down) &&
+                tr_variantDictFindInt(d, TR_KEY_rateUpload, &up) &&
+                tr_variantDictFindInt(d, TR_KEY_sizeWhenDone, &sizeWhenDone) &&
+                && tr_variantDictFindReal(d, TR_KEY_uploadRatio, &ratio))
+                */
+            bool include = negate;
+            std::string status = getStatusString(d);
+            std::string_view name;
```
Comment (ckerr) [inline utils/remote.cc:2029]:
```suggestion
```

File: utils/remote.cc [hunk 6]
```diff
@@ -1974,8 +1984,128 @@ static void printGroups(tr_variant* top)
...
+                        }
+                    }
+                }
+                break;
+            case 'n': // Torrent name substring
+                if (!tr_variantDictFindStrView(d, TR_KEY_name, &name))
+                {
+                    continue;
+                }
+                if (name.find(arg) != std::string::npos)
```
Comment (ckerr) [inline utils/remote.cc:2075]: scope of `name` can be narrowed

```suggestion
                if (std::string_view name; !tr_variantDictFindStrView(d, TR_KEY_name, &name))
                {
                    continue;
                }
                else if (name.find(arg) != std::string::npos)
```

**Comment by MatanZ** [comment]
I believe this addresses all your suggestions.

**Review by ckerr — APPROVED** [review]

**Comment by Coeur** [comment]
> I just noticed the naming in 'Files changed', is it correct? transmission-remote.1

Yes. Man pages have a numbered file extension.
- `man transmission-remote` will read `transmission-remote.1` since it's the only manual installed
- `man 2 transmission-remote` would read `transmission-remote.2` if it existed

Type `man man` for the manual of the `man` command.

**Pull request merged**